metadata-ingestion/src/datahub/ingestion/source/sql/data_reader.py

```diff
@@ -0,0 +1,137 @@
+                        - [<table_name>]
+            column: Column name
+        Returns:
+            list of column values
+        """
+
+        table = self._table(table_id)
+        query: Any
+        ignore_null_condition = sa.column(column_name).is_(None)
+        # limit doesn't compile properly for oracle so we will append rownum to query string later
+        if self.engine.dialect.name.lower() == "oracle":
+            raw_query = (
+                sa.select([sa.column(column_name)])
+                .select_from(table)
+                .where(sa.not_(ignore_null_condition))
+            )
+
+            query = str(
+                raw_query.compile(self.engine, compile_kwargs={"literal_binds": True})
+            )
+            query += "\nAND ROWNUM <= %d" % sample_size
+        else:
+            query = (
+                sa.select([sa.column(column_name)])
+                .select_from(table)
+                .where(sa.not_(ignore_null_condition))
+                .limit(sample_size)
+            )
+        query_results = self.engine.execute(query)
+
+        return [x[column_name] for x in query_results.fetchall()]
+
+    def get_sample_data_for_table(
+        self, table_id: List[str], sample_size: int = 100
+    ) -> Dict[str, list]:
+        """
+        Fetches table values, upto <sample_size>*1.2 count
+        Args:
+            table_id: Table name identifier. One of
+                        - [<db_name>, <schema_name>, <table_name>] or
+                        - [<schema_name>, <table_name>] or
+                        - [<table_name>]
+        Returns:
+            dictionary of (column name -> list of column values)
+        """
+        column_values: Dict[str, list] = defaultdict(list)
+        table = self._table(table_id)
+
+        # Ideally we do not want null values in sample data for a column.
+        # However that would require separate query per column and
+        # that would be expensiv. To compensate for possibility
+        # of some null values in collected sample, we fetch extra (20% more)
+        # rows than configured sample_size.
+        sample_size = int(sample_size * 1.2)
+
+        query: Any
+
+        # limit doesn't compile properly for oracle so we will append rownum to query string later
+        if self.engine.dialect.name.lower() == "oracle":
+            raw_query = sa.select([sa.text("*")]).select_from(table)
+
+            query = str(
+                raw_query.compile(self.engine, compile_kwargs={"literal_binds": True})
+            )
+            query += "\nAND ROWNUM <= %d" % sample_size
+        else:
+            query = sa.select([sa.text("*")]).select_from(table).limit(sample_size)
+        query_results = self.engine.execute(query)
+
+        # Not ideal - creates a parallel structure in column_values. Can we use pandas here ?
+        for row in query_results.fetchall():
+            if isinstance(row, LegacyRow):
```

Comment:
this LegacyRow thing seems fishy - is `row` always going to be a LegacyRow? won't it be a `Row` object in some cases?

we're on sqlalchemy 1.4, so LegacyRow should be deprecated https://docs.sqlalchemy.org/en/20/changelog/migration_14.html#rowproxy-is-no-longer-a-proxy-is-now-called-row-and-behaves-like-an-enhanced-named-tuple

Reply:
Agree. Given that LegacyRow import and functionality in sql_common.py is working fine, the current code flow should also work well for now. I've added fix for this in followup PR.